feat: Add ability to add IAM Auth to Lambda Function URLs

bestickley opened the pull request:
Fixes #108

This feature adds option to `NextjsDistribution` to enable IAM Auth to Lambda Function URLs (server and image optimization). Example usage:

```ts
new Nextjs(this, "NextSite", {
      nextjsPath: resolve(thisFilePath, "../../../../../ui"),
      defaults: {
        distribution: {
          functionUrlAuthType: FunctionUrlAuthType.AWS_IAM,
        }
      }
    });
```

References:
- https://medium.com/@dario_26152/restrict-access-to-lambda-functionurl-to-cloudfront-using-aws-iam-988583834705
- https://dev.to/aws-builders/signing-requests-with-aws-sdk-in-lambda-functions-476
- https://github.com/aws/aws-sdk-js/issues/86

revmischa commented on assets/lambda@edge/LambdaOriginRequestIamAuth.test.ts:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Simple test to debug lambda@edge functions
+ */
+describe('LambdaOriginRequestIamAuth', async () => {
```
this doesn't need to be async, you don't need to await test(...)

bestickley replied:
I was using the Node.js builtin test runner where you need to use `async` and `await` but now I see that this project is setup to use jest so I'll use jest and remove async/await

revmischa commented on assets/lambda@edge/LambdaOriginRequestIamAuth.test.ts:
```diff
@@ -0,0 +1,109 @@
+        },
+      ],
+    };
+    await handler(event, {} as CloudFrontRequestContext, () => {});
```
should be some expect call here

bestickley replied:
My original thought was to only use this for debugging purposes, but I'll add an expect call.

revmischa commented on assets/lambda@edge/LambdaOriginRequestIamAuth.ts:
```diff
@@ -0,0 +1,113 @@
+import { fixHostHeader, handleS3Request } from './common';
+
+/**
+ * This fixes the "host" header to be the host of the origin.
```
Comment should be updated

revmischa commented on assets/lambda@edge/LambdaOriginRequestIamAuth.ts:
```diff
@@ -0,0 +1,113 @@
+}
+
+function getSigV4(): SignatureV4 {
+  const region = process.env.AWS_REGION;
+  const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+  const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+  const sessionToken = process.env.AWS_SESSION_TOKEN;
```
i thought edge functions didn't have environment variables, are these automatically provided? i can't find docs on them

bestickley replied:
You're correct that environment variables cannot be used within Lambda@Edge functions. However, `process.env` is still injected with AWS credentials per request such that the AWS SDK's can be used within the functions to make API calls to AWS.

revmischa replied:
if it works then 👍🏻 
i just couldn't find any docs on what env vars edge functions have available

bestickley commented:
I fixed an issue where _next/image routes weren't working because the Origin Policy was preventing headers from getting through that were included in SigV4 IAM auth headers causing signature mismatch. I've updated signature calculation to restrict signature to these headers now so we're good to go.

Also, I saw build error because I updated jest configuration manually in package.json instead of through projen. I tried fixing it by updating in .projenrc.js file but it's not updating jest configuration :/ Tests weren't running on CI before so unless you want them running now I don't think there is any action required.

revmischa commented:
Thanks for your contribution! We're on the SST discord in #open-next if you want to chat further
